src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,143 @@
+## Specify contrasts for coding categorical data in model matrix. Contrast types
+## are constructed with data and other options, creating a subtype of
+## AbstractContrast. AbstractContrast provides the interface for creating modle
+## matrix columns and coefficient names
+##
+## ModelFrame will hold a Dict{Symbol, T<:AbstractContrast} that maps column
+## names to contrasts.
+##
+## ModelMatrix will check this dict when evaluating terms, falling back to a
+## default for any categorical data without a specified contrast.
+
+abstract AbstractContrast
+
+termnames(term::Symbol, col::Any, contrast::AbstractContrast) =
+    ["$term - $name" for name in contrast.termnames]
+
+function cols(v::PooledDataVector, contrast::AbstractContrast)
+    ## make sure the levels of the contrast matrix and the categorical data
+    ## are the same by constructing a re-indexing vector. Indexing into
+    ## reindex with v.refs will give the corresponding row number of the
+    ## contrast matrix
+    reindex = [findfirst(l .== contrast.levels) for l in levels(v)]
+
+    ## TODO: add kwarg for full-rank contrasts (e.g., when intercept isn't
+    ## specified in a model frame).
+
+    return contrast.matrix[reindex[v.refs], :]
+end
+
+## Default contrast is treatment coding:
+cols(v::PooledDataVector) = cols(v, TreatmentContrast(v))
+termnames(term::Symbol, col::PooledDataArray) = termnames(term, col, TreatmentContrast(col))
+
+
+## Constructing a contrast from a non-pooled data vector will first pool it
+## 
+## (NOT SURE THIS IS GOOD: constructing columns also depends on levels, so for
+## _that_ to work would need to somehow hold onto pooled data...)
```

Comment:
I'd thought about creating something like a `ContrastedPooledDataArray` like you describe, that wraps a PDA, but then looked at the `PooledDataArray` API and it seemed like more trouble than it was worth to create a wrapper type and delegate everything. Also, it seems like it might be prudent to see how categorical data implementation shakes out before marrying contrasts and categorical data itself, although I'm not sure what the status of that is at the moment (perhaps @johnmyleswhite wants to weigh in here...)